api/src/main/java/org/pmiops/workbench/workspaces/WorkspacesController.java

```diff
@@ -1010,6 +1044,76 @@ static String notebookLockingEmailHash(String bucket, String email) {
     return ResponseEntity.ok(recentWorkspaceResponse);
   }
 
+  private WorkspaceResource createWorkspaceResourceFromWorkspace(DbWorkspace workspace) {
+    return new WorkspaceResource()
+        .workspaceFirecloudName(workspace.getFirecloudName())
+        .workspaceId(workspace.getWorkspaceId())
+        .workspaceNamespace(workspace.getWorkspaceNamespace())
+        .workspaceBillingStatus(workspace.getBillingStatus());
+  }
+
+  @Override
+  public ResponseEntity<WorkspaceResourceResponse> getDataPageItems(
+      String workspaceNamespace, String workspaceId) {
+    // This also enforces registered auth domain.
+    workspaceService.enforceWorkspaceAccessLevel(
+        workspaceNamespace, workspaceId, WorkspaceAccessLevel.READER);
+
+    DbWorkspace workspace =
+        workspaceService.getRequiredWithCohorts(workspaceNamespace, workspaceId);
+    Set<DbCohort> cohorts = workspace.getCohorts();
```

Comment:
We can actually stuff all of the rest of this code into the mappers, since it's just a fetch & hydrate block. There's no real decision-making logic going on here.